[SPARK-38767][SQL] Support `ignoreCorruptFiles` and `ignoreMissingFiles` in Data Source options

Yaohua628 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
Support `ignoreCorruptFiles` and `ignoreMissingFiles` in Data Source options for both V1 and V2 file-based data sources.
```
spark.read...option("ignoreCorruptFiles", "true")...
spark.read...option("ignoreMissingFiles", "true")...
```


### Why are the changes needed?
Improve UX


### Does this PR introduce _any_ user-facing change?
Yes.

Previously:
```
spark.sql("set spark.sql.files.ignoreCorruptFiles=true")
spark.sql("set spark.sql.files.ignoreMissingFiles=true")
spark.read...
spark.read...
```

Now:
```
spark.read...option("ignoreCorruptFiles", "true")...
spark.read...option("ignoreMissingFiles", "true")...
```

### How was this patch tested?
Enhance existing UTs for ignoreMissingFiles + ignoreCorruptFiles

Yaohua628 committed:
draft for v1 and v2

Yaohua628 commented:
cc: @cloud-fan appreciate it if you can take a look whenever you have time, thanks!

Yaohua628 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVOptions.scala:
```diff
@@ -232,6 +232,12 @@ class CSVOptions(
   val unescapedQuoteHandling: UnescapedQuoteHandling = UnescapedQuoteHandling.valueOf(parameters
     .getOrElse("unescapedQuoteHandling", "STOP_AT_DELIMITER").toUpperCase(Locale.ROOT))
 
+  val ignoreCorruptFiles: Boolean = parameters.get("ignoreCorruptFiles").map(_.toBoolean)
```
Maybe we could have a common trait for those file-based options (CSVOptions, JSONOptions, etc), and pass it to `FilePartitionReaderFactory `, `FilePartitionReader `, WDYT?

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SourceOptions.scala:
```diff
@@ -47,4 +47,6 @@ object SourceOptions {
   val RESPECT_SPARK_SCHEMA = "respectSparkSchema"
   val DEFAULT_RESPECT_SPARK_SCHEMA = false
 
```
maybe all the `XXXOptions` should extend `SourceOptions`, so that we can put common code here, e.g.
```
val ignoreCorruptFiles: Boolean = parameters.get("ignoreCorruptFiles").map(_.toBoolean)
    .getOrElse(SQLConf.get.ignoreCorruptFiles)
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVPartitionReaderFactory.scala:
```diff
@@ -45,7 +45,10 @@ case class CSVPartitionReaderFactory(
+    filters: Seq[Filter])
+  extends FilePartitionReaderFactory(
+    parsedOptions.ignoreCorruptFiles,
+    parsedOptions.ignoreMissingFiles) {
```
We can avoid this change, if `CSVOptions` extends `SourceOptions`, and `FilePartitionReaderFactory` simply get these two flags from the options.

Yaohua628 committed:
refactor

Yaohua628 committed:
new options

Yaohua628 committed:
some nits

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReaderFactory.scala:
```diff
@@ -16,20 +16,21 @@
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
-abstract class FilePartitionReaderFactory extends PartitionReaderFactory {
+abstract class FilePartitionReaderFactory(options: FileSourceOptions)
```
A simpler is to add it as a method
```
abstract class FilePartitionReaderFactory extends PartitionReaderFactory {
  protected def options: FileSourceOptions
  ...
}
```

cloud-fan replied:
then we don't need to change its subclasses, as they are case class and already implicitly provided the `options` method.

Yaohua628 replied:
wow! thanks, that's simpler!

Yaohua628 committed:
comments and ignoreMissingFiles test

Yaohua628 committed:
nit

Yaohua628 commented:
@cloud-fan Addressed your comments, thanks for that! also, add a test in `FileBasedDataSourceSuite`.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReaderFactory.scala:
```diff
@@ -16,20 +16,23 @@
 
 abstract class FilePartitionReaderFactory extends PartitionReaderFactory {
+
+  protected def parsedOptions: FileSourceOptions
```
For this new method, can we use simply `options` instead of `parsedOptions`?

cloud-fan replied:
`parsedOptions` is the existing name, renaming it will need to change all the subclasses of `FilePartitionReaderFactory`. How about we do it in a separated PR?

cloud-fan replied:
ah sorry, `parsedOptions` is not that widely used, let's rename it to `options` then.

Yaohua628 replied:
sure - done!

dongjoon-hyun reviewed (commented):
Thank you, @Yaohua628 and @cloud-fan .
- I updated the `Affected Version` of this improvement JIRA to 3.4.0 from 3.2.1. 
- I left one minor comment for the naming about `parsedOptions`.

HyukjinKwon reviewed (commented):
we would have to document it at https://github.com/apache/spark/blob/master/docs/sql-data-sources-generic-options.md

Yaohua628 committed:
rename to options and adding doc + examples

Yaohua628 commented:
> Thank you, @Yaohua628 and @cloud-fan .
> 
> * I updated the `Affected Version` of this improvement JIRA to 3.4.0 from 3.2.1.
> * I left one minor comment for the naming about `parsedOptions`.

Thanks for fixing `Affected Version`! Also, renamed to `options`.


> we would have to document it at https://github.com/apache/spark/blob/master/docs/sql-data-sources-generic-options.md

Thanks! Added doc and updated examples.

cc: @cloud-fan

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you for revising.

dongjoon-hyun commented on examples/src/main/python/sql/datasource.py:
```diff
@@ -28,12 +28,25 @@
     # dir1/file3.json is corrupt from parquet's view
-    test_corrupt_df = spark.read.parquet("examples/src/main/resources/dir1/",
+    test_corrupt_df1 = spark.read.parquet("examples/src/main/resources/dir1/",
                                          "examples/src/main/resources/dir1/dir2/")
```
Hi, @Yaohua628 . This seems to break the Python Linter CI.
```
./examples/src/main/python/sql/datasource.py:35:17: E128 continuation line under-indented for visual indent
[21](https://github.com/Yaohua628/spark/runs/5967096763?check_suite_focus=true#step:15:21)
                "examples/src/main/resources/dir1/dir2/")
[22](https://github.com/Yaohua628/spark/runs/5967096763?check_suite_focus=true#step:15:22)
                ^
[23](https://github.com/Yaohua628/spark/runs/5967096763?check_suite_focus=true#step:15:23)
./examples/src/main/python/sql/datasource.py:48:42: E128 continuation line under-indented for visual indent
[24](https://github.com/Yaohua628/spark/runs/5967096763?check_suite_focus=true#step:15:24)
                                         "examples/src/main/resources/dir1/dir2/")
```

Yaohua628 replied:
ah yeah, thanks! fixed it

dongjoon-hyun replied:
Oh, it seems that you are hitting GitHub Action outage.

Yaohua628 replied:
Just tried now, and it is still failing cc: @cloud-fan 
```
Ref: spark-38767
SHA: c02d34370cc71312cf396aaf6550700e5f022cce
Error: There was a new unsynced commit pushed. Please retrigger the workflow.
Error: Unhandled error: Error: There was a new unsynced commit pushed. Please retrigger the workflow.
    at eval (eval at callAsyncFunction (/home/runner/work/_actions/actions/github-script/f05a81df23035049204b043b50c3322045ce7eb3/dist/index.js:6[120](https://github.com/apache/spark/runs/5984516854?check_suite_focus=true#step:2:120):56), <anonymous>:72:11)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async main (/home/runner/work/_actions/actions/github-script/f05a81df23035049204b043b50c3322045ce7eb3/dist/index.js:6147:20)
```

cloud-fan replied:
We can check the GA test manually in https://github.com/Yaohua628/spark/actions/workflows/build_and_test.yml

Yaohua628 replied:
Thanks! Seems like all tests passed: https://github.com/Yaohua628/spark/actions/runs/2152938291

dongjoon-hyun replied:
Thank you for confirmation, @Yaohua628 .

Yaohua628 committed:
fix Python linter

Yaohua628 committed:
empty commit: re-trigger tests

Yaohua628 committed:
Merge branch 'master' into spark-38767

Yaohua628 committed:
empty commit: re-trigger tests

Yaohua628 committed:
Merge branch 'master' into spark-38767

dongjoon-hyun commented:
Merged to master for Apache Spark 3.4.0.
Thank you, @Yaohua628 , @cloud-fan , @HyukjinKwon

dongjoon-hyun commented:
Welcome to the Apache Spark community, @Yaohua628 .
I added you to the Apache Spark contributor group and SPARK-38767 is assigned to you.

Yaohua628 commented:
@dongjoon-hyun Thank you!! 🎉
(btw I think you assigned to the wrong person "Yaohua Cui", should be "Yaohua Zhao", :-P. I know, it's confusing)

dongjoon-hyun commented:
Oh.. Sorry. Now I fixed it. BTW, this is the previous my worklog. Although I chose `1` for `Yaohua Zhao` and it showed you correctly in the log, it seems that Apache Spark script had some bug for those two IDs. Maybe, I can take a look at that later. BTW, thank you for informing that.
```
JIRA is unassigned, choose assignee
[1] Yaohua Zhao (Reporter)
Enter number of user, or userid, to assign to (blank to leave unassigned):1
=== JIRA SPARK-38767 ===
summary		Support ignoreCorruptFiles and ignoreMissingFiles in Data Source options
assignee	Yaohua Zhao
status		In Progress
url		https://issues.apache.org/jira/browse/SPARK-38767

Enter comma-separated fix version(s) [3.4.0]:
Successfully resolved SPARK-38767 with fixVersions=['3.4.0']!
```

Yaohua628 commented:
> Oh.. Sorry. Now I fixed it. BTW, the following is the previous my worklog. Although I chose `1` for `Yaohua Zhao` and it showed you correctly in the log, it seems that Apache Spark script had some bug for those two IDs. Maybe, I can take a look at that later. BTW, thank you for informing that.
> 
> ```
> JIRA is unassigned, choose assignee
> [1] Yaohua Zhao (Reporter)
> Enter number of user, or userid, to assign to (blank to leave unassigned):1
> === JIRA SPARK-38767 ===
> summary		Support ignoreCorruptFiles and ignoreMissingFiles in Data Source options
> assignee	Yaohua Zhao
> status		In Progress
> url		https://issues.apache.org/jira/browse/SPARK-38767
> 
> Enter comma-separated fix version(s) [3.4.0]:
> Successfully resolved SPARK-38767 with fixVersions=['3.4.0']!
> ```

Ah got it! Not a problem. Thank you so much for the help on this PR!!

Yaohua628 commented:
> it seems that Apache Spark script had some bug for those two IDs.

might relate to this one, a JIRA python client bug: https://github.com/pycontribs/jira/issues/1049

Yaohua628 commented:
@dongjoon-hyun hi Dongjoon, just realize I probably need to fix [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala#L155) by passing the `options` in `InMemoryFileIndex`, where `ignoreMissingFiles` is used, right? Please correct me if I misunderstand this part! Thanks a lot!

Yaohua628 commented:
@cloud-fan @dongjoon-hyun Hi folks, do you know when the doc changes will be deployed? https://spark.apache.org/docs/latest/sql-data-sources-generic-options.html? Thanks!

dongjoon-hyun commented:
@Yaohua628  As I commented https://github.com/apache/spark/pull/36069#issuecomment-1097149876 , this is Apache Spark 3.4.0 for next year. What you are referring is Apache Spark 3.3, isn't it? 

For Apache Spark 3.4.0, we just started to discuss about the release plan in the dev mailing list.
- https://lists.apache.org/thread/0dt1n6rhhhvsyj48y8bfq3sxvxvlpn57

dongjoon-hyun commented:
FYI, Apache Spark 3.4 code freeze was March 15th, 2022. This patch simply arrived one month later after code freeze deadline.

- https://spark.apache.org/versioning-policy.html

<img width="829" alt="Screen Shot 2022-07-20 at 8 28 13 PM" src="https://user-images.githubusercontent.com/9700541/180123691-02c55883-f163-4c26-a976-6108fb464c4f.png">

Yaohua628 commented:
@dongjoon-hyun Oh yeah, you are right! Thanks for your clarification!
